vumi/middleware/session_length.py

```diff
@@ -42,37 +42,43 @@ def setup_middleware(self):
     def teardown_middleware(self):
         yield self.redis.close_manager()
 
+    def _generate_redis_key(self, message, address):
+        return '%s:%s:%s' % (
+            message.get('transport_name'), address, 'session_created')
+
+    @inlineCallbacks
+    def _set_redis_time(self, redis_key, time):
+        if (yield self.redis.get(redis_key)) is None:
+            yield self.redis.setex(redis_key,  self.timeout, str(time))
+
     def _set_message_session_metadata(self, message, field, value):
-        if not message['helper_metadata'].get(self.field_name):
-            message['helper_metadata'][self.field_name] = {}
-        message['helper_metadata'][self.field_name][field] = value
+        metadata = message['helper_metadata'].setdefault(self.field_name, {})
+        metadata.setdefault(field, value)
 
     def _set_session_start_time(self, message, time):
+        time = float(time)
         self._set_message_session_metadata(message, 'session_start', time)
 
     def _set_session_end_time(self, message, time):
+        time = float(time)
         self._set_message_session_metadata(message, 'session_end', time)
 
     @inlineCallbacks
     def _process_message(self, message, redis_key):
         if message.get('session_event') == self.SESSION_NEW:
             start_time = self.clock.seconds()
-            yield self.redis.setex(redis_key,  self.timeout, str(start_time))
+            yield self._set_redis_time(redis_key, start_time)
```

Comment:
>  Justin van der Merwe [4:53 PM] 
>  hodgestar: should we change the middleware to not overwrite values but still overwrite the redis value, or should it overwrite neither?
>
>  Justin van der Merwe [4:54 PM]
>  overwriting neither seems sensible
>
>  Simon Cross [4:54 PM] 
>  +1

I guess you understood my suggestion as meaning "overwrite both the redis value and the message start time value only if the message has no start time value".

I'm wondering if we would ever need to explicitly set message start times somewhere else (without changing the redis value), but I guess we can wait for that use case to appear before trying to support that.

I can't think of a way that the two values could get out of sync, but I'm sure are ways I'm not thinking of. +1 on the suggested approach